### PR TITLE
Create a member list of the cluster using etcd

### PR DESCRIPTION
**Pull request by hackerwins**
**What this PR does / why we need it**:

Create a member list of the cluster using etcd
- [x] Fix to be able to launch multiple Agents at the same time.
- [x] Implement distributed PubSub without broadcasting.
- [x] Create a member list of the cluster using etcd.

Screen Shot 2021-05-12 at 7 37 34 PM

**Which issue(s) this PR fixes**:

Address #183

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation**:

```docs

```

**Checklist**:
- [x] Added relevant tests or not required
- [x] Didn't break anything

### CODE REVIEW

File: yorkie/backend/sync/etcd/membermap.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
+package etcd
```
Comment (dc7303) [inline yorkie/backend/sync/etcd/membermap.go:1]: It looks like you need to add Copyright.

**Review by dc7303 — COMMENTED** [review]
Thanks for the wonderful work.
I left a few comments while reviewing. Please check.

**Review by dc7303 — COMMENTED** [review]
If you run the entire test several times, there are cases of intermittent failures. Although there are many types of errors, it appears to be an issue occurring in an asynchronous execution environment. The log attached below is one of the errors that have occurred.

I think it is good to solve the intermittent errors that occur in asynchronous situation in this PR. What do you think?
```bash
--- FAIL: TestClusterMode (2.15s)
    --- FAIL: TestClusterMode/member_list_test (1.08s)
        cluster_mode_test.go:45: 
                Error Trace:    cluster_mode_test.go:45
                Error:          Not equal: 
                                expected: map[string]*sync.AgentInfo{"/agents/c2h7dcabtmil1tvdo6m0":(*sync.AgentInfo)(0xc000558690), "/agents/c2h7dcibtmil1tvdo6mg":(*sync.AgentInfo)(0xc00045e370), "/agents/c2h7dcibtmil1tvdo6n0":(*sync.AgentInfo)(0xc00018eff0)}
                                actual  : map[string]*sync.AgentInfo{"/agents/c2h7dcabtmil1tvdo6m0":(*sync.AgentInfo)(0xc0005583c0), "/agents/c2h7dcibtmil1tvdo6mg":(*sync.AgentInfo)(0xc000558460), "/agents/c2h7dcibtmil1tvdo6n0":(*sync.AgentInfo)(0xc000558500)}
                                
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -472,3 +472,3 @@
                                   UpdatedAt: (time.Time) {
                                -   wall: (uint64) 570157000,
                                +   wall: (uint64) 562257000,
                                    ext: (int64) 63756856754,
                Test:           TestClusterMode/member_list_tes
```

File: yorkie/backend/backend.go [hunk 2]
```diff
@@ -65,31 +69,48 @@ func New(
...
 	var lockerMap sync.LockerMap
 	if etcdConf != nil {
-		etcdClient, err := etcd.Dial(etcdConf)
+		etcdClient, err := etcd.Dial(etcdConf, agentInfo)
 		if err != nil {
 			return nil, err
 		}
+		if err := etcdClient.Initialize(); err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (hackerwins) [reply yorkie/backend/backend.go:102]: I don't think it's a problem if agents receive broadcast as the subscriptions are not created on the RPC. What specific problems do you expect?

File: yorkie/backend/sync/etcd/membermap.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+// setAgentInfo sets the given agentInfo to the local member map.
+func (c *Client) setAgentInfo(key string, value sync.AgentInfo) {
+	c.memberMapMu.Lock()
+	defer c.memberMapMu.Unlock()
+
+	c.memberMap[key] = &value
+}
+
```
[the remark this answers is not in the document]
Comment (hackerwins) [reply yorkie/backend/sync/etcd/membermap.go:159]: I think the problem will be solved if we duplicate the member list and return it.

**Comment by hackerwins** [comment]
I repeated the test about 50 times to reproduce the first problem, but I couldn't. How did you reproduce it?

```
for i in {1..50}; do make test; done
```

**Comment by dc7303** [comment]
> I repeated the test about 50 times to reproduce the first problem, but I couldn't. How did you reproduce it?
> 
> ```
> for i in {1..50}; do make test; done
> ```

I noticed that it occurs intermittently when running the test with the mentioned command. The way to reproduce was to just run the test.

In addition, a case of failure due to the different number of members was confirmed.
```bash
--- FAIL: TestClusterMode (2.14s)
    --- FAIL: TestClusterMode/member_list_test (1.05s)
        cluster_mode_test.go:53: 
                Error Trace:    cluster_mode_test.go:53
                Error:          "map[c2icvpabtmik0ttu96hg:%!s(*sync.AgentInfo=&{c2icvpabtmik0ttu96hg choedongcheol-ui-iMac.local localhost:1101 {309873000 63757010661 0x2a20680}}) c2icvpabtmik0ttu96i0:%!s(*sync.AgentInfo=&{c2icvpabtmik0ttu96i0 choedongcheol-ui-iMac.local localhost:2101 {901292000 63757010661 0x2a20680}}) c2icvpabtmik0ttu96ig:%!s(*sync.AgentInfo=&{c2icvpabtmik0ttu96ig choedongcheol-ui-iMac.local localhost:3101 {923449000 63757010661 0x2a20680}}) c2icvpibtmik11jl324g:%!s(*sync.AgentInfo=&{c2icvpibtmik11jl324g choedongcheol-ui-iMac.local localhost:1201 {913513000 63757010662 0x2a20680}})]" should have 3 item(s), but has 4
                Test:           TestClusterMode/member_list_test
        cluster_mode_test.go:56: 
                Error Trace:    cluster_mode_test.go:56
                Error:          "map[c2icvpabtmik0ttu96hg:%!s(*sync.AgentInfo=&{c2icvpabtmik0ttu96hg choedongcheol-ui-iMac.local localhost:1101 {309873000 63757010661 0x2a20680}}) c2icvpabtmik0ttu96ig:%!s(*sync.AgentInfo=&{c2icvpabtmik0ttu96ig choedongcheol-ui-iMac.local localhost:3101 {923449000 63757010661 0x2a20680}}) c2icvpibtmik11jl324g:%!s(*sync.AgentInfo=&{c2icvpibtmik11jl324g choedongcheol-ui-iMac.local localhost:1201 {913513000 63757010662 0x2a20680}})]" should have 2 item(s), but has 3
                Test:           TestClusterMode/member_list_test
        cluster_mode_test.go:59: 
                Error Trace:    cluster_mode_test.go:59
                Error:          "map[c2icvpabtmik0ttu96hg:%!s(*sync.AgentInfo=&{c2icvpabtmik0ttu96hg choedongcheol-ui-iMac.local localhost:1101 {309873000 63757010661 0x2a20680}}) c2icvpibtmik11jl324g:%!s(*sync.AgentInfo=&{c2icvpibtmik11jl324g choedongcheol-ui-iMac.local localhost:1201 {913513000 63757010662 0x2a20680}})]" should have 1 item(s), but has 2
                Test:           TestClusterMode/member_list_test

```

**Comment by dc7303** [comment]
It was confirmed that the current `member list test` test code was modified. This seems to solve the problem of `UpdatedAt` mismatch in the test. 👍 

But I've analyzed this problem and found that there are some things we need to consider. We'll have to discuss whether this should be solved in this thread, but I hope we're aware of it.

**Red box**: call `putAgent()` from AgentB (3101 port)
**Blue box**: memory address of AgentB

Screen Shot 2021-05-20 at 1 41 39 AM

```go
// yorkie/backend/syncetcd/membermap.go
func (c *Client) Initialize() error {
	ctx := context.Background()
    // # 1
	if err := c.putAgent(ctx); err != nil {
		return err
	}
    // # 2
	if err := c.initializeMemberMap(ctx); err != nil {
		return err
	}
    
    // # 3
	go c.syncAgents()
    // # 4
	go c.putAgentPeriodically()

	return nil
}
```

1. AgentB information is registered in # 1 line, and agentB information is updated in the rest of agentA and defaultYorkie.
2. In # 2, agentB stores other agent information and its own information.
3. Activate the Watch mode in # 3.
4. In #4, putAgent is called again, but agentB calls the Members() method in the test before executing `setAgentInfo()`.

The conclusion seems to be an issue caused by calling Members() from another task in a situation where the event did not reach the specific agent being watched.

In addition to UpdatedAt, I think there may be a moment when the agent information known to each other in the actual operating environment may be different.

File: yorkie/backend/backend.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hackerwins) [reply yorkie/backend/backend.go:102]: Client RPC and RPC for broadcast will be separated. We need to consider the problem situation mentioned when implementing RPC for broadcast.

**Comment by hackerwins** [comment]
@dc7303 Thank you for the comment.
The comment you gave is characteristic of optimistic replication. We can make the mistake of assuming that these traits are exactly the same, so we need to be careful when implementing features.

**Review by dc7303 — APPROVED** [review]
LGTM 👍

**Pull request merged**